### ISSUE TITLE
[MNG-5756] Java home output in mvn -v is misleading

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#7672]
**[Jarkko Rantavuori](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jarkkor)** opened **[MNG-5756](https://issues.apache.org/jira/browse/MNG-5756?redirect=false)** and commented

For example on my windows box, mvn -v prints the following:

```
Java home: C:\Program Files (x86)\Java\jdk1.7.0_51\jre
```

But my JAVA_HOME is actually

```
> echo %JAVA_HOME%
C:\Program Files (x86)\Java\jdk1.7.0_51
```

In the source code, the line comes from:

https://git-wip-us.apache.org/repos/asf?p=maven.git;a=blob;f=maven-embedder/src/main/java/org/apache/maven/cli/CLIReportingUtils.java#l63

```
version.append( "Java home: " ).append( System.getProperty( "java.home", "<unknown java home>" ) ).append( ls );
```

which is using property "java.home" to fetch java home. However, "java.home" property is not JAVA_HOME! This is explained in detail in here: http://javahowto.blogspot.fi/2006/05/javahome-vs-javahome.html

To quote:

> What's the difference between JAVA_HOME and java.home?
> 
> JAVA_HOME is the JDK install directory, e.g., C:\jdk5. It's meant to be set as an environment variable and referenced in Windows batch files or Unix scripts. I always have it in my Windows Control Panel and .tcsh files,along with other common environment variables. Some Java applications use the name jdk.home for this purpose, which I think is a better name. But JAVA_HOME has been used since the beginning and is now a convention.
> 
> java.home is the JRE install directory, e.g., C:\jdk5\jre, or C:\Program Files\Java\jre1.5.0_06. Unlike JAVA_HOME, I never seen java.home as an environment variable. java.home is a build-in Java system property, whose value is the JRE install directory. Since all Java system properties are also exposed as Ant build properties, you can also use $&lcub;java.home&rcub; in build files.
> 
> Would jre.home be a better name? Maybe, but I don't think Sun will change it.

This is a source of constant confusion. Some stackoverflow threads to illustrate:

http://stackoverflow.com/questions/15279586/java-home-in-maven
http://stackoverflow.com/questions/17620531/maven-pointing-to-jre-instead-of-jdk

The correct way to print JAVA_HOME would be to use System.getenv("JAVA_HOME"). Either that should be used or current output should be changed so it wouldn't be so misleading.

---

**Affects:** 3.2.5, 3.3.3

**Remote Links:**
- [GitHub Pull Request #115
](https://github.com/apache/maven/pull/115)
- [GitHub Pull Request #117
](https://github.com/apache/maven/pull/117)
- [GitHub Pull Request #117
](https://github.com/apache/maven/pull/117)

1 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

See this reference, a Java app runs always on top of a JRE: https://docs.oracle.com/javase/tutorial/essential/environment/sysprop.html

I might be confusing but the output is correct. Does it matter for you to the the environment variable `JAVA_HOME` instead of `java.home`?

**Comment by jira-importer** [comment]
**[Jarkko Rantavuori](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jarkkor)** commented

It does matter, when users constantly confuse it with JAVA_HOME (like I pointed out in stackoverflow threads. there are others as well). JAVA_HOME is the thing everybody knows and can set directly, so when Maven output is talking about "Java home", it is understandable that users get confused. In common language, JAVA_HOME is indeed the thing everybody in java world means when talking about java home. Furthermore, Maven startup scripts use JAVA_HOME to look up which java they use for running Maven, so it is actually the relevant piece of information when configuring Maven. I can't imagine a reason why mvn -v output would show jre used for running.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I do understand your point but Maven isn't showing you the JRE home but the VM does. It is a mere property.

What one could do in 3.4 is to show:

```
Java home: .../java8 JRE home: .../java8/jre
```

**Comment by jira-importer** [comment]
**[Jarkko Rantavuori](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jarkkor)** commented

Agreed. But it is Maven talking about "Java home", and it is maven doing

```
System.getProperty( "java.home", "<unknown java home>" ) 
```

instead of

```
System.getenv("JAVA_HOME")
```

when describing it.

Your suggestion sounds good to me. There might be some confusion if some user thinks "JRE home" is something one should set in the environment, like JAVA_HOME, so personally I would maybe say something like

```
Java home: .../java8
JRE used: .../java8/jre
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

`System.getProperty( "java.home", "<unknown java home>" )` this is correct. Though it is confusing. You should complain to Sun/Oracle. `JAVA_HOME` != `java.home`. Though I have not tried with `JAVA_HOME` set to a JRE.

**Comment by jira-importer** [comment]
**[Jarkko Rantavuori](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jarkkor)** commented

> System.getProperty( "java.home", "\<unknown java home>" ) this is correct. Though it is confusing.

Well, I don't agree on the correctness. Maven shouldn't be saying things about "java home" or "Java home" when it is actually referring to "java.home". I don't think you disagree about the statement that normally when anybody in java ecosystem is talking about java home, they refer to JAVA_HOME environment setting. Given that, if output of a program is meaning something else with it, it would be the fault of that program - in this case Maven. This has been evident with tens of users tripping on this on stackoverflow.

> JAVA_HOME != java.home.

Agreed. And `java.home` != `java home`. But "java home", if that term is to be used, really should refer to JAVA_HOME.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Just a note: with Java 9 this will be completely irrelevant, see [slides](http://openjdk.java.net/projects/jigsaw/talks/prepare-for-jdk9-j1-2015.pdf) 29, 30. You won't see any difference.

**Comment by jira-importer** [comment]
**[Jarkko Rantavuori](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jarkkor)** commented

Surely this isn't close-pending due to that? Java 9 GA is currently scheduled for March 2017, and has already been delayed. It's hardly a reason to not fix this now. Even when it hits GA this will continue to be a problem in earlier versions.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I do agree. Though, there is little participation on this issue and not further complaints about confusion.

**Comment by jira-importer** [comment]
**[Jarkko Rantavuori](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jarkkor)** commented

I can provide a patch/PR if that's what's needed. I guess this is too trivial to interest people in participation, but in SO there's already been ~20 votes about the subject.

**Comment by jira-importer** [comment]
**[Marcelo Souza](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=marcelosv)** commented

Hi. I committed but my e-mail have not correct. My email out marcelo@ceossistemas.com but the correct is marcelosouzav@gmail.com.

I change in my git for correct email.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

So I pushed a test branch for this. The solution is a reduced version of both PRs because I see no reason to print out `JAVA_HOME` any more. From Java 9 it will correspond to `java.home`. The output does not say "Java home" anymore, but solely references "JRE used". In fact, you don't even need to set `JAVA_HOME` because your default installation in `/usr/bin/` or `/usr/local/bin` shall be enough.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [9abfc88425601da4e756a80768c2215bb97e9725](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=9abfc88425601da4e756a80768c2215bb97e9725).
